dm-lwm2m: ZmP v38 changes

mike-scott opened the pull request:
- Correct OpenThread / 802.15.4 check in Kconfig
- Add OpenThread overlay
- Enable GPIO explicitly now as some boards have had many defaults removed

mike-scott committed:
dm-lwm2m: fix OpenThread / 802.15.4 check

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
dm-lwm2m: add OpenThread configuration changes

This can be used for OpenThread support with the following boards:
- nrf52840_pca10056
- reel_board

Support for nrf52840_pca10059 requires some debugging due to
it's default configuration in Zephyr.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
dm-lwm2m: enable GPIO config for sample

Upstream Zephyr has reduced the # of configs enabled by default.
This includes CONFIG_GPIO.  We need to explicitly select this
as our sample uses GPIOs for LEDs, etc.

Signed-off-by: Michael Scott <mike@foundries.io>

mbolivar commented on overlay-ot.conf:
```diff
@@ -0,0 +1,49 @@
+# Hack: Disable BT on devices which support both for now
```
Can we just change the application Kconfig to say:


```
config BT
	default y if !NET_L2_OPENTHREAD
```

?

mike-scott replied:
This is actually an open item for Nordic to fix. 
See zephyr/boards/arm/nrf52840_pca10056/nrf52840_pca10056_defconfig line 22
CONFIG_BT=y

mbolivar replied:
Got it, thanks.

mbolivar commented on overlay-ot.conf:
```diff
@@ -0,0 +1,49 @@
+CONFIG_BT=n
+
+CONFIG_CPLUSPLUS=y
+CONFIG_FLASH=y
```
Why are some options given in this overlay with the same values they have in prj.conf? That shouldn't be necessary.

For options which need different options when using IPSP vs. OpenThread, does it make sense to handle in the Kconfig using default values to have a single point of truth instead of relying on kconfig to override prj.conf with overlay-ot.conf values? Seems a little nicer to me and may result in a smaller overlay file which is easier to read.

mike-scott replied:
I went through and removed the duplicates.  I was using the same overlay file for testing the upstream LwM2M client.

mbolivar commented on overlay-ot.conf:
```diff
@@ -0,0 +1,49 @@
+CONFIG_NET_L2_OPENTHREAD=y
+
+CONFIG_NET_CONFIG_MY_IPV6_ADDR=""
+CONFIG_NET_CONFIG_PEER_IPV6_ADDR="64:ff9b::0527:53ce"
```
What is this? Could use a comment.

mbolivar commented on overlay-ot.conf:
```diff
@@ -0,0 +1,49 @@
+CONFIG_NET_IF_MCAST_IPV6_ADDR_COUNT=8
+
+#CONFIG_OPENTHREAD_DEBUG=y
+#CONFIG_NET_LOG=y
```
Same here, we've already got a debug helpers section in prj.conf which has some of these (like CONFIG_NET_LOG), why duplicate this?

mike-scott commented:
Addressed comments (thanks!)

mbolivar commented on overlay-ot.conf:
```diff
@@ -0,0 +1,43 @@
+CONFIG_FLASH_PAGE_LAYOUT=y
+
+# Disable Proxy
+CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_COAP_PROXY_SUPPORT=n
```
I still think it would make sense to not override values here that are set in prj.conf if it's possible to set them using defaults in the Kconfig file instead. The .conf files will be smaller and have less duplicated information.

For example, this option is default n, so it seems like it could be enabled for BT only with this line in Kconfig:

```
config LWM2M_FIRMWARE_UPDATE_PULL_COAP_PROXY_SUPPORT
	default y if BT
```

Then it could be removed from prj.conf, overlay-modem.conf, and this file.

I've got similar suggestions elsewhere in this file. What do you think?

mbolivar commented on overlay-ot.conf:
```diff
@@ -0,0 +1,43 @@
+CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_COAP_PROXY_SUPPORT=n
+
+# Disable IPv4
+CONFIG_NET_IPV4=n
```
Since this option is default n already, it seems like this line could only be needed for Kinetis K6X, since there's a 'select NET_IPV4' line under FOTA_DEVICE_SOC_SERIES_KINETIS_K6X.

If that's right, can't this line be deleted and just make FOTA_DEVICE_SOC_SERIES_KINETIS_K6X `depends on !NET_L2_OPENTHREAD`? It seems like K6X won't run the OT configuration no matter what, might as well express the dependency in the Kconfig if so.

mbolivar commented on overlay-ot.conf:
```diff
@@ -0,0 +1,43 @@
+CONFIG_NET_IPV4=n
+CONFIG_NET_CONFIG_NEED_IPV4=n
+CONFIG_NET_CONFIG_MY_IPV4_ADDR=""
+CONFIG_NET_CONFIG_PEER_IPV4_ADDR=""
```
The above three IPV4 lines would be deletable as well.

mbolivar commented on overlay-ot.conf:
```diff
@@ -0,0 +1,43 @@
+CONFIG_MAIN_STACK_SIZE=2048
+CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=2048
+CONFIG_NET_TX_STACK_SIZE=2048
+CONFIG_NET_RX_STACK_SIZE=5120
```
Same here, what do you think about

```
config NET_RX_STACK_SIZE
	default 5120 if NET_L2_OPENTHREAD
	default 2048
```

mbolivar commented on overlay-ot.conf:
```diff
@@ -0,0 +1,43 @@
+CONFIG_NET_RX_STACK_SIZE=5120
+
+# Disable conflicting IEEE802154 L2
+CONFIG_NET_L2_IEEE802154=n
```
This option already depends on `!NET_L2_OPENTHREAD`; are this and the next line really needed? Seems like they could be removed.

mbolivar commented on overlay-ot.conf:
```diff
@@ -0,0 +1,43 @@
+CONFIG_NET_L2_OPENTHREAD=y
+
+# OpenThread will set local IP
+CONFIG_NET_CONFIG_MY_IPV6_ADDR=""
```
Why is this needed? The option has no defaults, so it implicitly defaults to the empty string.

mbolivar commented on overlay-ot.conf:
```diff
@@ -0,0 +1,43 @@
+CONFIG_NET_CONFIG_PEER_IPV6_ADDR="64:ff9b::c6c7:6c9f"
+
+# Thread by default registers quite a lot addresses.
+CONFIG_NET_IF_UNICAST_IPV6_ADDR_COUNT=6
```
Similarly, for these, what about something like this?

```
config NET_IF_UNICAST_IPV6_ADDR_COUNT
        default 3 if BT
        default 6 if NET_L2_OPENTHREAD

config NET_IF_MCAST_IPV6_ADDR_COUNT
        default 2 if BT
        default 8 if NET_L2_OPENTHREAD
```

mbolivar reviewed (commented):
Thanks for the updates! Some more detailed comments on my idea about making the overlays smaller.

mike-scott commented:
@mbolivar Thank you for clarifying, I sort of misunderstood your initial discussion.  Let me take a stab at cleaning this up.

mike-scott committed:
dm-lwm2m: Kconfig: K64F doesn't support OpenThread

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
dm-lwm2m: Kconfig: default FOTA_LED_GPIO_INVERTED y for SOC_NRF52840

All of the existing nrf52840 SoC HW uses this flag.  Let's set it
in Kconfig and remove the individual settings per board.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
dm-lwm2m: Kconfig: default LwM2M proxy setting to y for BT

Instead of individual config settings in conf files enabling /
disabling the PROXY setting, let's establish that it's only used
when BT is enabled on nRF52 devices.

NOTE: the generic BT settings may be moved later so that they
are not specific to nRF52 devices.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
dm-lwm2m: Kconfig: move ADDR_COUNT settings out from BT section

Let's establish a general defaults section and move the ADDR_COUNT
settings there with "if BT" clauses for the current defaults.

This allows us to later add other defaults for these settings.

No effective .config changes with this move.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
dm-lwm2m: Kconfig: set default for NET_RX_STACK_SIZE

Later, we can set the default net RX stack size without needing to
add a .conf setting.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
Updated to include Kconfig / .conf cleanup prior to adding overlay-ot.conf

I compared the .configs between old and new for all of our current targets and there weren't any changes.

mike-scott commented:
I will be adding the ones that apply to Hawkbit/MQTT there as well.

mike-scott commented on Kconfig:
```diff
@@ -26,7 +26,7 @@ if FOTA_DEVICE_SOC_SERIES_KINETIS_K6X
 # IPv6 is enabled by default in Zephyr's Kconfig when networking is selected.
 config NET_IPV6
-	default n
+	default n if FOTA_DEVICE_SOC_SERIES_KINETIS_K6X
```
Woops.  Just realized this is already in side a FOTA_DEVICE_SOC_SERIES_KINETIS_K6X block

mbolivar reviewed (approved):
Nice and clean! Thanks a lot for making these changes; the ot overlay is super legible now.
